test/XmlHttpRequestConnection.tests.js

```diff
@@ -145,6 +145,80 @@ define(function (require) {
 
         });
 
+        describe("is correctly using XmlHttpRequest and runs attached callbacks while performing:", function (){
+            beforeEach(function (){
+                mockXMLHttpRequest.prototype.send = function (body) {
+                    this.readyState = 4;
+                    this.status = 200;
+                    this.onloadstart();
+                    this.onload();
+                    this.onloadend();
+                    this.onprogress();
+                    this.ontimeout();
+                    this.onerror();
+                    this.onabort();
+                    this.upload.onloadstart();
+                    this.upload.onload();
+                    this.upload.onloadend();
+                    this.upload.onprogress();
+                    this.upload.ontimeout();
+                    this.upload.onerror();
+                    this.upload.onabort();
+                    this.onreadystatechange();
+                };
+                mockXMLHttpRequest.prototype.upload = {};
+                spyOn(mockXMLHttpRequest.prototype, 'send').andCallThrough();
+
+                window.XMLHttpRequest = (function () { return mockXMLHttpRequest; }());
+
+                connection = new XmlHttpRequestConnection();
+            });
+
+            it("execute call and invokes custom request event callbacks", function () {
```

Comment:
> Well, so you are mocking up the XHR object and in fact it might look like you are testing the XHR object itself

actually much less than in your patch where you manually call all the `on*` event handlers and where you set a `readyState` and `status`.

>  It should be a simple mock

Given, XmlHttpRequestConnection builds an XmlHttpRequest instance we can not use an actual mock but by using `addEventListener` we would be able to spy `XmlHttpRequest.prototype.addEventListener` instead of completely replace `window.XmlHttpRequest`.

> IMO the solution you proposed is overcomplicated and it doesn't bring much value.

it brings the value of clearly showing which event handlers are supported or not. It also bring the value of having **unit** tests, the unit being the support for **each** event handler separately.

>  Everytime we do a request we are setting up a new XHR object hence why I don't see a point in using addEventListener methods in this case when you can apply one callback to one event.

the point: ease unit tests. But it's also possible to properly test the code where you set `on*` event handlers. Both are fine provided it's correctly unit tested.

> The other thing is we are not reusing XHR objects. They will be garbage collected when they finish handling request.

I don't see the relation with the `on*` event handlers vs. `addEventListener` call.

Anyway, as written above, both ways are fine provided it's correctly unit tested, so feel free to choose what you prefer, as long as it's correctly unit tested that's fine.